src/replication/packing_stream.cpp

```diff
@@ -0,0 +1,647 @@
+    {
+      return error_code;
+    }
+
+  m_data_start_position = serializator->get_stream_read_position ();
+  /* force read (fetch data from stream source) */
+  stream_start_ptr = serializator->extend_unpacking_range (get_data_packed_size (), &m_buffered_range);
+  if (stream_start_ptr == NULL)
+    {
+      error_code = ER_FAILED;
+      return error_code;
+    }
+
+  /* the stream entry header is unpacked, and its contents are ready to be unpacked */
+
+  return error_code;
+}
+
+/* this is called only right before applying the replication data */
+int stream_entry::unpack (void)
+{
+  int i, error_code = NO_ERROR;
+
+  /* TODO[arnia] : make sure the serializator range already points to data contents */
+
+  stream_packer *serializator = get_packer ();
+  size_t count_packable_entries = get_packable_entry_count_from_header ();
+  BUFFER_UNIT *stream_start_ptr = serializator->start_unpacking_range_from_pos (m_data_start_position,
+                                                                                get_data_packed_size (),
+                                                                                &m_buffered_range);
+
+  for (i = 0 ; i < count_packable_entries; i++)
+    {
+      /* create a specific replication_entry object (depending on type) */
+      packable_object *packable_entry = dynamic_cast<packable_object *>(get_builder()->create_object (serializator));
+      
+      assert (packable_entry != NULL);
+
+      /* unpack one replication entry */
+      add_packable_entry (packable_entry);
+      packable_entry->unpack (serializator);
+    }
+ 
+  return error_code;
+}
+
+int stream_entry::add_packable_entry (packable_object *entry)
+{
+  m_packable_entries.push_back (entry);
+
+  return NO_ERROR;
+}
+
+size_t stream_entry::get_entries_size (void)
+{
+  size_t entry_size, total_size = 0;
+  int i;
+
+  stream_packer *serializator = get_packer ();
+  for (i = 0; i < m_packable_entries.size (); i++)
+    {
+      entry_size = m_packable_entries[i]->get_packed_size (serializator);
+      entry_size = DB_ALIGN (entry_size, MAX_ALIGNMENT);
+      total_size += entry_size;
+    }
+
+  return total_size;
+}
+
+
+////////////////////////////////////////
+
+packing_stream::packing_stream (const buffer_provider *my_provider)
+{
+  if (my_provider == NULL)
+    {
+      m_buffer_provider = buffer_provider::get_default_instance ();
+    }
+  else
+    {
+      m_buffer_provider = (buffer_provider *) my_provider;
+    }
+  m_last_reported_ready_pos = 0;
+  m_max_buffered_position = 0;
+  m_total_buffered_size = 0;
+  m_read_position = 0;
+
+  /* TODO[arnia] : system parameter */
+  trigger_flush_to_disk_size = 1024 * 1024;
+
+  set_filled_stream_handler (NULL);
+  set_fetch_data_handler (NULL);
+}
+
+int packing_stream::init (const stream_position &start_position)
+{
+  m_append_position = start_position;
+
+  return NO_ERROR;
+}
+
+int packing_stream::write (const size_t &byte_count, stream_handler *handler)
```

Comment:
do we really need const size_t & instead of size_t? by using const size_t & it can be possible to hinder the compiler's optimization strategies.